src/world.mojo

```diff
@@ -0,0 +1,1705 @@
+        # self._filter_cache:    newCache(),
+
+        # var node = self.createArchetypeNode(Mask, ID, false)
+        # self.createArchetype(node, Entity, false)
+
+    fn _get_archetype_index(
+        inout self,
+        components: InlineArray[ComponentInfo, _],
+        start_node_index: Int = 0,
+    ) -> Int:
+        """Returns the archetype list index of the archetype differing from
+        the archetype at the start node by the given indices.
+
+        If not start node is given, returns the archetypes with the
+        components at the given indices.
+
+        If necessary, creates a new archetype.
+
+        Args:
+            components:       The components that distinguish the archetypes.
+            start_node_index: The index of the start archetype's node.
+
+        Returns:
+            The archetype list index of the archetype differing from the start
+            archetype by the components at the given indices.
+        """
+        node_index = self._archetype_map.get_node_index(components)
+        if self._archetype_map.has_value(node_index):
+            return self._archetype_map[node_index]
+
+        archetype_index = len(self._archetypes)
+        self._archetype_map[node_index] = archetype_index
+
+        self._archetypes.append(Archetype(node_index, components))
+
+        return archetype_index
+
+    fn new_entity(inout self) raises -> Entity as entity:
+        """Returns a new or recycled [Entity].
+
+        Raises when called on a locked world.
+        Do not use during [Query] iteration!
+        """
+        self._check_locked()
+        entity = self._create_entity(0)
+
+    fn new_entity[
+        *Ts: ComponentType
+    ](inout self, *components: *Ts) raises -> Entity as entity:
+        """Returns a new or recycled [Entity].
+        The given component types are added to the entity.
+
+        Raises when called on a locked world.
+        Do not use during [Query] iteration!
+
+        ⚠️ Important:
+        Entities are intended to be stored and passed around via copy, not via pointers! See [Entity].
+
+        Parameters:
+            Ts: The components to add to the entity.
+
+        Args:
+            components: The components to add to the entity.
+        """
+        self._check_locked()
+
+        alias size = components.__len__()
+
+        component_info = self._component_manager.get_info_arr[*Ts]()
+        archetype_index = self._get_archetype_index(component_info)
+        entity = self._create_entity(archetype_index)
+        index_in_archetype = self._entities[int(entity.id)].index
+
+        archetype = self._archetypes.get_ptr(archetype_index)
+
+        @parameter
+        for i in range(components.__len__()):
+            archetype[].unsafe_set(
+                int(index_in_archetype),
+                component_info[i].id,
+                UnsafePointer.address_of(components[i]).bitcast[UInt8](),
+            )
+
+        # TODO
+        # if self._listener != nil:
```

Comment:
If I guessed right, the listeners are something for the observer pattern. I would like to implement this, too.

Reply:
No, you would implement observers (for model output) as systems.

Listeners send events on "ECS operations", namely entity creation and removal, as well as component addition and removal. This allows you to react to operations done by certain systems, but in a different place.

An example use case would be death of entities possible in several places/systems, with some cleanup required. You would create a system (or it could also be a resource) that subscribes to the respective events (e.g removal of entities with a certain [combination of] component[s]) and does the cleanup.

You could also handle this differenty, e.g. by calling the cleanup in every place where you kill the respective entities. But that would introduce dependencies ("killer" system needs to know the cleanup requirement), and would be error-prone.